### DISCUSSION TITLE
Opentelemetry java auto-instrumentation session tracing

### DISCUSSION THREAD

**Discussion by NamitaJaokar** [open-telemetry/opentelemetry-java-instrumentation#5547] [Q&A]
Hi,

I am using opentelemetry javaagent jar version 10 and currently trying to implement session tracing using opentelemetry with zipkin as exporter.
Is there any configurations available to trace or export the users authenticated in the application ?
Also, I have tried few configuration properties at my end to uniquely define the users if at all I am able to capture hostname/IP address of the users using the agent.
-Dotel.metrics.exporter : This property , gives me the hostname of the machine and and other OS and process level when the value is set to logging. But the same fails when tried with zipkin as the metrics exporter. Is metrics export option available only with logging and not Zipkin exporter?

Apart from the above points , I also wanted to understand below configuration properties. I had tried setting these properties in my application but there were no changes/additions in the kind of attributes/traces captured before and after setting them
-Dotel.instrumentation.http.capture-headers.client.request
-Dotel.instrumentation.http.capture-headers.client.response
-Dotel.instrumentation.http.capture-headers.server.request
-Dotel.instrumentation.http.capture-headers.server.response
-Dotel.instrumentation.common.peer-service-mapping
For all of the above properties overridden , I had set the value as true for them. 

Can someone please advise/ help me on the above queries.

Thanks in advance

**Comment by trask** [comment]
hi @NamitaJaokar! 

> But the same fails when tried with zipkin as the metrics exporter

Zipkin is not a metrics backend, you may want to look at prometheus if you are looking for an OSS metrics backend.

> For all of the above properties overridden , I had set the value as true for them.

these properties are not boolean valued, check out https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/v1.10.0/docs/config/common.md

**Reply by NamitaJaokar** [reply]
Hi @trask ,

Thank you for the references provided. I have tried adding few of them to the properties but I cannot find them as attributes in the traces.
For example, I tried giving below values for the respective properties:
-Dotel.instrumentation.http.capture-headers.server.request=http.client_ip, http.route
-Dotel.instrumentation.common.peer-service-mapping= 123.456.7.890=my_service_name
Could you please correct me if at all the values or syntax does not match the required syntax or values.

Thank you

**Reply by trask** [reply]
> otel.instrumentation.http.capture-headers.server.request=http.client_ip, http.route

these should be the http header names that you want to capture

**Reply by NamitaJaokar** [reply]
Hi @trask , 

Is there any example of the above that I can refer to. Also,  are there any configurations available in Opentelemetry to add custom attributes? If yes , Can you give some suggestions /advise to configure the same.

Thanks in advance.

**Reply by trask** [reply]
can you describe in detail what you want to capture? that will help us point you in the right direction

**Reply by NamitaJaokar** [reply]
I am looking out for 2 things 
1. Capture http header attributes like authorization details.
2.  Some Custom attributes in the traces or data collected and exported to the exporter ( in my case its zipkin exporter ). For example application.users=some user or list of users.

**Reply by NamitaJaokar** [reply]
@trask 

Can you help me with  understanding and configuring below attributes
![image](https://user-images.githubusercontent.com/32196871/158120976-374cfc71-136d-4dab-aa23-51c927d74386.png)

I came across these while going through below link:
https://github.com/open-telemetry/opentelemetry-specification/blob/main/specification/trace/semantic_conventions/span-general.md

Thank you

**Reply by trask** [reply]
hi @NamitaJaokar, there's no auto-instrumentation for `enduser.*` attributes, so you would need to capture those manually, e.g.

Span.current().setAttribute("enduser.id", "myuser")

**Comment by mvkrishna86** [comment]
Hi @trask , Is it possible to capture the headers (without any processing) with the auto-instrumentation? 
Like OTEL_PYTHON_DJANGO_TRACED_REQUEST_ATTRS for Django, do we have any option for Java?

**Reply by trask** [reply]
Is this what you're looking for? https://github.com/open-telemetry/opentelemetry-java-instrumentation/blob/main/docs/config/common.md#capturing-http-request-and-response-headers

**Reply by mvkrishna86** [reply]
Thanks @trask . This is exactly what I am looking for.
